Fixed EditorProperties resetting when being edited.

EricEzaM opened the pull request:
Fixes #41809. Thank @Paulb23 for the prompt on what the issue was.

After fixing the Editor Inspector autorefresh (#41697), every EditorProperty which used a LineEdit or TextEdit to display and alter it's value was calling `set_text()` every refresh. `set_text()` includes `cursor_pos = 0;` which is what caused this issue. To circumvent this, `EditorProperty`'s will only update if they are _not_ selected in the inspector. This means they will not refresh while being edited.

Paulb23 commented:
Nice work!

There's also a secondary issue with `Array` and `Dictionary` properties not just for text input, as they seem to be completely reload from scratch on refresh.

EricEzaM commented:
@Paulb23 Ok, due to that I had a rethink of the solution and it is now much simpler. The refresh will just not occur if the property is selected. This solves all issues because the refresh was just immediately overwriting all the changes the user had made.

Paulb23 commented:
Took a quick look. `Arrays` and `Dictionarys` `key/value` editors, aren't count as selected, so causes now this behaviour:

![array_edit_bug_2](https://user-images.githubusercontent.com/6584330/93522878-0ec5ad00-f92a-11ea-9596-59520ee528b6.gif)

Guess the easiest way is to select the parent element automatically, though nested `Arrays/Dictionarys` might be a problem.

EricEzaM commented:
Huh, that's weird, when I tested arrays and dicts were actually selected when being edited. Damn.

EricEzaM commented:
@Paulb23 Ok, I think I finally got it. I needed to add some logic so that the selected status is passed to the EditorPropertyArray/Dictionary when one of it's child properties are selected or focused.

A simple Inspector Refresh sure has wreaked some havoc.

Paulb23 commented:
Awesome, yeah it's always the simplest changes. 

Gave it a quick local test, looks good for the most part, only one issue that I can find is with the colour picker:

![color_picker_bug](https://user-images.githubusercontent.com/6584330/93640126-f6b96080-f9f1-11ea-9860-44180181b253.gif)

EricEzaM committed:
Fixed editor inspector properties not being able to be edited due to Inspector Auto-refresh overriding the changes.

Inspector will not refresh any properties who are a parent of the current GUI focus owner.

EricEzaM commented:
@Paulb23 Ok, I think I am back to a simple fix which actually works. The property will just not update if the GUI focus owner is one of it's children. This should account for all input's which could be affected by this issue, including ones in custom EditorProperties in EditorPlugins.

I did try this earlier but for some reason thought it didn't work - but I was mistaken, it does work.

Paulb23 commented:
No worries, this does seem to work okay, just need to check if `get_focus_owner()` returns `null` else you get: `ERROR: Parameter "p_node" is null. at: is_a_parent_of (scene/main/node.cpp:1461)`. Then it looks good to me.

EricEzaM commented:
Done, I am now using `windows_subsystem=console` in my debug builds so hopefully I pick up on some of these errors which I did not see before.

Paulb23 reviewed (approved):
LGTM.

akien-mga commented on editor/editor_inspector.cpp:
```diff
@@ -2364,6 +2364,11 @@ void EditorInspector::_notification(int p_what) {
 			if (refresh_countdown <= 0) {
 				for (Map<StringName, List<EditorProperty *>>::Element *F = editor_property_map.front(); F; F = F->next()) {
 					for (List<EditorProperty *>::Element *E = F->get().front(); E; E = E->next()) {
+						// Don't update the property if one of it's children is currently focused.
```
```suggestion
						// Don't update the property if one of its children is currently focused.
```

akien-mga commented on editor/editor_inspector.cpp:
```diff
@@ -2364,6 +2364,11 @@ void EditorInspector::_notification(int p_what) {
 				for (Map<StringName, List<EditorProperty *>>::Element *F = editor_property_map.front(); F; F = F->next()) {
 					for (List<EditorProperty *>::Element *E = F->get().front(); E; E = E->next()) {
+						// Don't update the property if one of it's children is currently focused.
+						// This prevents the property being editor from being overriden by the current value in update_property().
```
```suggestion
						// This prevents the property being edited from being overriden by the current value in update_property().
```

akien-mga commented:
I'm not sure about this fix, should we address the regression instead of working it around?

It seems the bug was introduced by #41697. Is it that the inspector was never refreshing before? Or less frequently?

EricEzaM commented:
@akien-mga, before that fix the inspector was never refreshing at all. The logic was completely disconnected.

I am not sure how else to fix this. The `update_property() ` methods seem to all call some version of `set_text`, `set_value`, etc, which also updates the editor property display. When that is being called every 0.3 seconds it makes it impossible to edit values in the inspector as they are being reset before you can change them.

akien-mga commented:
I guess we just need to check with @reduz whether the previous fix was correct, or whether force updating the inspector every 0.3 second is a good idea or not. If it was implemented but never worked, maybe it's not meant to be actually used.

akien-mga commented:
Closing as it was decided that the proper fix is to revert #41697, see #42106.
